core/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java

```diff
@@ -448,7 +453,21 @@ public void finalizeRecovery(final long targetLocalCheckpoint) {
         StopWatch stopWatch = new StopWatch().start();
         logger.trace("finalizing recovery");
         cancellableThreads.execute(() -> {
-            shard.markAllocationIdAsInSync(request.targetAllocationId(), targetLocalCheckpoint);
+            /*
+             * Before marking the shard as in-sync we acquire an operation permit. We do this so that there is a barrier between marking a
+             * shard as in-sync and relocating a shard. If we acquire the permit then no relocation handoff can complete before we are done
+             * marking the shard as in-sync. If the relocation handoff holds all the permits then after the handoff completes and we acquire
+             * the permit then the state of the shard will be relocated and this recovery will fail.
+             */
+            final PlainActionFuture<Releasable> onAcquired = new PlainActionFuture<>();
+            shard.acquirePrimaryOperationPermit(onAcquired, ThreadPool.Names.SAME);
```

Comment:
👍 I wrote a long paragraph explaining why we would need to do this as I missed this somehow in my initial pass. 🤣 

I would like to have an assertion that no more updates to the GlobalCheckPointTracker are happening once the PrimaryContext is sampled and successfully communicated to the primary relocation target. For this, we could introduce a `sealed` boolean to `GlobalCheckPointTracker`. When sampling the `PrimaryContext`, this is set to `true`. If there are any failures then later in the `relocated`, we could unseal the `GlobalCheckPointTracker` again (can be solved nicely with a Releaseable that is returned by primaryContext sampling method). We could then add checks on every update method of `GlobalCheckPointTracker` that it is not sealed.